src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php

```diff
@@ -127,6 +127,9 @@ protected function getConstructor(Definition $definition, $required)
         }
 
         if ($factory) {
+            if ($factory instanceof Reference) {
```

Comment:
Even though it's best to do the parsing into `[Reference, '__invoke']` at the loader level, would it be useful to leave this here so that factories can be defined like `$definition->setFactory(new Reference($id))` in PHP service configuration files?

Or assume that PHP service configuration files must be explicit in setting invokable factories?

I'm not sure where else to put this functionality if not in the pass.

Reply:
maybe we can do this in `Definition::setFactory()`?

Reply:
That's a much better idea!